### PR TITLE
Add sync support

### PR DESCRIPTION
**Pull request by arv**
Towards https://github.com/rocicorp/replicache/issues/30

### CODE REVIEW

**Comment by arv** [comment]
I don't like the exposure of beginSync and maybeEndSync. Even with all those annotations they "show up" in some places.

Maybe remove the test once things get more stable.

**Comment by arv** [comment]
This does not yet do the scheduling of syncing every X seconds. I will deal with that later.

File: test/replicache_test.dart [hunk 1]
```diff
@@ -38,14 +78,38 @@ Future<void> main() async {
...
+  }
+
   const MethodChannel(CHANNEL_NAME)
       .setMockMethodCallHandler((methodCall) async {
     final method = methodCall.method;
     final String dbName = methodCall.arguments[0];
     final Uint8List data = methodCall.arguments[1];
+
+    if (recordPath == null && replays.isNotEmpty) {
+      final i = replays.indexWhere((r) => r.matches(dbName, method, data));
```
Comment (aboodman) [inline test/replicache_test.dart:100]: [taste] Is it possible to just play the responses in order, one after the next, instead of searching for a match?

File: test/replicache_test.dart [hunk 2]
```diff
@@ -38,14 +78,38 @@ Future<void> main() async {
 
   // TODO(arv): Start the test server from here!
 
+  List replays = [];
+  String recordPath;
+
+  Future<void> useReplay(String path) async {
```
Comment (aboodman) [inline test/replicache_test.dart:84]: [help] Why have the replay mechanism? I thought that part of the goal was to test the entire stack, including replicache-client, so I'm surprised to see this here.

File: test/replicache_test.dart (hunk 1 above)
Comment (arv) [reply test/replicache_test.dart:100]: Not at the moment since 'pull' comes along at "random" times. I'm tempted to expose `Replicache.forTesting` which would allow disabling sync and access `maybeEndSync` etc.

File: test/replicache_test.dart (hunk 2 above)
Comment (arv) [reply test/replicache_test.dart:84]: I added it so that we can run the tests without working on the live data. Will be useful for CI. I'm fine removing this for now if we do not think it is worth keeping?

File: test/replicache_test.dart (hunk 1 above)
Comment (aboodman) [reply test/replicache_test.dart:100]: I think it would be useful to disable sync anyway, so you could make that a feature. `Replicache.syncInterval=null`

File: lib/replicache.dart [hunk 3]
```diff
@@ -109,15 +113,14 @@ class Replicache implements ReadTransaction {
...
   /// Create or open a local Replicache database with named `name` synchronizing with `remote`.
   /// If `name` is omitted, it defaults to `remote`.
-  Replicache(this._remote, {String name = "", String clientViewAuth = ""})
-      : _clientViewAuth = clientViewAuth {
-    if (_remote == "") {
-      throw new Exception("remote must be non-empty");
-    }
-    if (name == "") {
-      name = _remote;
+  Replicache(this._remote,
```
[the remark this answers is not in the document]
Comment (arv) [reply lib/replicache.dart]: This is actually diff server pull url. This needs some cleaning up once things work.

File: test/replicache_test.dart (hunk 1 above)
Comment (arv) [reply test/replicache_test.dart:100]: Another reason to have forTesting is to allow awaiting for `_opened`.

Comment (arv) [reply test/replicache_test.dart:100]: Still. I agree. We should allow pausing the sync interval

File: lib/replicache.dart [hunk 4]
```diff
@@ -192,8 +230,102 @@ class Replicache implements ReadTransaction {
...
+  Future<void> _maybeEndSync(String syncHead) async {
+    await _opened;
+    if (_closed) {
+      return;
+    }
+    final res = await _invoke(_name, 'maybeEndSync', {'syncHead': syncHead});
+    final replayMutations = res['replayMutations'];
+    if (replayMutations == null || replayMutations.isEmpty) {
+      // All done.
+      await _checkChange(syncHead);
```
Comment (phritz) [inline lib/replicache.dart:293]: super newb to this code but this might be a bug. we expect root to be equal to synchead instantly after maybeendsync succeeds, so even though head changed this call to checkChange will not fire onchange.

File: lib/replicache.dart [hunk 5]
```diff
@@ -192,8 +230,102 @@ class Replicache implements ReadTransaction {
...
+        'basis': basis,
+        'original': original,
+      }
+    });
+    return res.ref;
+  }
+
+  /// Synchronizes the database with the server. New local transactions that
+  /// have been executed since the last sync are sent to the server, and new
+  /// remote transactions are received and replayed.
```
Comment (phritz) [inline lib/replicache.dart:328]: i would not say that new remote transactions are received and replayed. i would say that new state from the client view is received and any local mutations that are not included in that client view are replayed.

File: lib/replicache.dart [hunk 6]
```diff
@@ -211,36 +343,13 @@ class Replicache implements ReadTransaction {
...
-          _reauthenticating = false;
-        } else {
-          await _checkChange(result['root']);
-          break;
-        }
-      }
+      await _sync();
       _scheduleSync(5);
     } catch (e) {
       // We are seeing some consistency errors during sync -- we push commits,
```
Comment (phritz) [inline lib/replicache.dart:349]: I don't understand this comment, probably is outdated. We do *not* necessarily expect that if we successfully push commit C and then pull a new client view that it includes C. We do not require the batch endpoint to operate synchronously or for it to be consistent with the clientview.

File: lib/replicache.dart [hunk 7]
```diff
@@ -192,8 +230,102 @@ class Replicache implements ReadTransaction {
...
+  }
+
+  Future<void> _maybeEndSync(String syncHead) async {
+    await _opened;
+    if (_closed) {
+      return;
+    }
+    final res = await _invoke(_name, 'maybeEndSync', {'syncHead': syncHead});
+    final replayMutations = res['replayMutations'];
+    if (replayMutations == null || replayMutations.isEmpty) {
```
Comment (phritz) [inline lib/replicache.dart:291]: if replay mutations is null/empty that is not sufficient to say that sync completed successfully. we also need to check that maybeendsync did not return an error because replaymutations will be empty in that case too. i dont know dart and dont understand how the error comes back from https://github.com/rocicorp/replicache-client/blob/8a2d797818bdb97b6ebbd0e2538d167c5408562b/repm/api.go#L181 so maybe this is handled in some way i dont see.

File: lib/replicache.dart [hunk 8]
```diff
@@ -192,8 +230,102 @@ class Replicache implements ReadTransaction {
...
       query((tx) => tx.scan(prefix: prefix, start: start, limit: limit));
 
-  /// Synchronizes the database with the server. New local transactions that have been executed since the last
-  /// sync are sent to the server, and new remote transactions are received and replayed.
+  Future<void> _sync() async {
+    await _opened;
+    if (_closed) {
+      return;
+    }
+    final syncHead = await _beginSync();
```
Comment (phritz) [inline lib/replicache.dart:238]: what happens if beginsync errors?

File: lib/replicache.dart (hunk 4 above)
Comment (aboodman) [reply lib/replicache.dart:293]: I think it's right.

The bindings don't monitor the database directly, the only way they know that the master changed is when either tx commits or sync lands, and tells it that master (may have) changed.

Comment (aboodman) [reply lib/replicache.dart:293]: See: https://github.com/rocicorp/replicache-sdk-flutter/blob/master/lib/replicache.dart#L277

File: lib/replicache.dart (hunk 5 above)
Comment (aboodman) [reply lib/replicache.dart:328]: This was my original phrasing, and I agree with your critique. Let me try again... these comments are worth fretting over since they will show up in e.g., autocomplete when people are using Replicache.

> Synchronizes this cache with the server. New local mutations are sent to the server, and the latest server state is applied to the cache. Any local mutations not included in the new server state are replayed. See the Replicache design document for more information on sync: https://github.com/rocicorp/replicache/blob/master/design.md

File: lib/replicache.dart (hunk 6 above)
Comment (aboodman) [reply lib/replicache.dart:349]: Good point - @arv we can remove this entire catch block actually.

File: lib/replicache.dart (hunk 7 above)
Comment (aboodman) [reply lib/replicache.dart:291]: The errors from Go get translated into exceptions by lower-level machinery. `sync` catches any exceptions at all and re-schedules itself.

File: lib/replicache.dart (hunk 4 above)
Comment (phritz) [reply lib/replicache.dart:293]: hmm, i still don't understand. on line 293 we are passing synchead to checkchange. since sync just completed we expect synchead to be equal to head/root so line 277 will be false and it will not fire onchange. what's the intent of doing a check like that?

Comment (arv) [reply lib/replicache.dart:293]: There is an issue here. We should not check for changes until the end of sync. Right now we check for the intermediate commits.

File: lib/replicache.dart (hunk 8 above)
Comment (arv) [reply lib/replicache.dart:238]: Throws an exception

File: lib/replicache.dart [hunk 9]
```diff
@@ -192,8 +230,102 @@ class Replicache implements ReadTransaction {
...
+      final httpStatusCode = data['httpStatusCode'];
+      switch (httpStatusCode) {
+        case HttpStatus.ok:
+          break;
+        case HttpStatus.unauthorized:
+          // TODO: Reauth
+          break;
+        default:
+          print(
+              'Got unexpected response from $serverName server: $httpStatusCode: ${data['errorMessage']}');
```
[the remark this answers is not in the document]
Comment (arv) [reply lib/replicache.dart:263]: "Got errror response from..."
